Coverage for instance type resource using linter mnd

anishbista60 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
there was a need to exclude some instancetype source files for the 'mnd' linter.
After this PR:
Remove the instancetype source files from the 'mnd' ignored files by solving the errors they produce.
<!-- (optional, in `fixes #<issue number>(, fixes #12662 , ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #12662 

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- https://github.com/kubevirt/kubevirt/issues/12662 -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

anishbista60 commented:
@EdDev works fine in my machine . Could you please take a look ?

orelmisan reviewed (commented):
Thank you for the PR @anishbista60 
Please see the inline comments.

anishbista60 commented on tests/instancetype/instancetype.go:
```diff
@@ -1706,9 +1716,9 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									CPU: &virtv1.CPU{
-										Cores:   uint32(2),
+										Cores:   uint32(MemorySize2),
 										Threads: uint32(1),
-										Sockets: uint32(2),
+										Sockets: uint32(MemorySize2),
```
@orelmisan should i revert the changes for use other name .

anishbista60 commented on tests/instancetype/instancetype.go:
```diff
@@ -401,7 +411,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 			var apiStatus errors.APIStatus
 			Expect(goerrors.As(err, &apiStatus)).To(BeTrue(), "error should be type APIStatus")
-			Expect(apiStatus.Status().Details.Causes).To(HaveLen(3))
+			Expect(apiStatus.Status().Details.Causes).To(HaveLen(expectedCausesLength))
```
@orelmisan do you mean like this . 

```
expectedcauses := []metav1.StatusCause{
				{
					Type:    "",
					Message: ",
					Field:   "",
				},
			}
			Expect(apiStatus.Status().Details.Causes).To(HaveLen(expectedCausesLength))
			Expect(apiStatus.Status().Details.Causes).To(Equal(expectedcauses))
```
if yes , what to put into type, message and field . Could you please help me here ?

anishbista60 replied:
@0xFelix here could you please review

anishbista60 replied:
> @0xFelix here could you please review?

lyarwood replied:
That's already done below but could be refactored in a follow up, I wouldn't change it as part of this PR.

anishbista60 replied:
@lyarwood ok sir

0xFelix commented on tests/instancetype/instancetype.go:
```diff
@@ -47,6 +47,19 @@ import (
 
 var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-compute] Instancetype and Preferences", decorators.SigCompute, func() {
 	var virtClient kubecli.KubevirtClient
+	const (
+		preferredTerminationGracePeriodSeconds = 15
+		expectedCausesLength                   = 3
+		timeout                                = 300 * time.Second
+		dvSuccessTimeoutSeconds                = 180
+		guestMemorySize                        = 2
+		numCore                                = 2
+		numSocket                              = 2
+		numThread                              = 2
+		MemorySize4                            = 4
+		memorySize2GB                          = 2 * 1024 * 1024 * 1024
+		memorySize1GB                          = 1 * 1024 * 1024 * 1024
+	)
```
Are all these constants really necessary? I would rather only add constants for the places that gomnd is complaining about.

anishbista60 replied:
@0xFelix  ok three constant are added additionally , I'll remove that

anishbista60 commented:
@0xFelix could you please take look ? Thank you

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -47,6 +47,17 @@ import (
+		MemorySize4                            = 4
+		memorySize2GB                          = 2 * 1024 * 1024 * 1024
+		memorySize1GB                          = 1 * 1024 * 1024 * 1024
+	)
```
Can you please move most of these constants down into their relevant contexts?

lyarwood commented on pkg/instancetype/instancetype.go:
```diff
@@ -802,7 +803,7 @@ func (m *InstancetypeMethods) InferDefaultInstancetype(vm *virtv1.VirtualMachine
 		var ignoreableInferenceErr *IgnoreableInferenceError
 		if errors.As(err, &ignoreableInferenceErr) && ignoreFailure {
 			//nolint:gomnd
```
Can you also drop this?

lyarwood commented on pkg/instancetype/instancetype.go:
```diff
@@ -838,7 +839,7 @@ func (m *InstancetypeMethods) InferDefaultPreference(vm *virtv1.VirtualMachine)
 
 		if errors.As(err, &ignoreableInferenceErr) && ignoreFailure {
 			//nolint:gomnd
```
As above.

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1426,6 +1433,13 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
+			memorySize2GB   = 2 * 1024 * 1024 * 1024
+			memorySize1GB   = 1 * 1024 * 1024 * 1024
+			guestMemorySize = 2
+			numCPU          = 2
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1449,7 +1463,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 					Spec: instancetypev1beta1.VirtualMachineInstancetypeSpec{
 						CPU: instancetypev1beta1.CPUInstancetype{
-							Guest: uint32(2),
+							Guest: uint32(guestMemorySize),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1460,7 +1474,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1537,7 +1551,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`requiredNumvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1557,7 +1571,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 										Cores:   uint32(1),
 										Threads: uint32(1),
-										Sockets: uint32(2),
+										Sockets: uint32(numCPU),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1577,7 +1591,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1595,7 +1609,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									CPU: &virtv1.CPU{
-										Cores:   uint32(2),
+										Cores:   uint32(numCPU),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1648,7 +1662,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1667,7 +1681,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 									CPU: &virtv1.CPU{
 										Cores:   uint32(1),
-										Threads: uint32(2),
+										Threads: uint32(numCPU),
```
`providedvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1688,7 +1702,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(4),
+								Guest: uint32(MemorySize4),
```
`requiredvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1706,9 +1720,9 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									CPU: &virtv1.CPU{
-										Cores:   uint32(2),
+										Cores:   uint32(numCPU),
```
Can you create new constants for these as total vCPU count is sockets * cores * threads.

Something like `providedCores` and `providedSockets`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1788,7 +1802,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`requiredvCPUs` ?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1867,7 +1881,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`requiredvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1908,7 +1922,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`requiredvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1949,7 +1963,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(2),
+								Guest: uint32(guestMemorySize),
```
`requiredvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1990,7 +2004,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 						Requirements: &instancetypev1beta1.PreferenceRequirements{
 							CPU: &instancetypev1beta1.CPUPreferenceRequirement{
-								Guest: uint32(4),
+								Guest: uint32(MemorySize4),
```
`requiredvCPUs`?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -2008,7 +2022,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									CPU: &virtv1.CPU{
-										Cores:   uint32(2),
+										Cores:   uint32(numCPU),
```
`providedvCPUs` ?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -2046,7 +2060,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									Memory: &virtv1.Memory{
-										Guest: resource.NewQuantity(1*1024*1024*1024, resource.BinarySI),
+										Guest: resource.NewQuantity(memorySize1GB, resource.BinarySI),
```
Can you just change this into a var?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1743,7 +1757,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									Memory: &virtv1.Memory{
-										Guest: resource.NewQuantity(2*1024*1024*1024, resource.BinarySI),
+										Guest: resource.NewQuantity(memorySize2GB, resource.BinarySI),
```
Can you just change this into a var?

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -47,6 +47,9 @@ import (
 var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-compute] Instancetype and Preferences", decorators.SigCompute, func() {
 	var virtClient kubecli.KubevirtClient
+	const (
+		timeout = 300 * time.Second
```
`timeoutSeconds`?

Also can be a single line above `var`?

lyarwood replied:
This should just be:

```
const timeoutSeconds =  300 * time.Second
```

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1426,6 +1433,13 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
+		const (
+			memorySize2GB   = 2 * 1024 * 1024 * 1024
+			memorySize1GB   = 1 * 1024 * 1024 * 1024
+			guestMemorySize = 2
```
This doesn't need to be defined.

anishbista60 commented:
@lyarwood Done sir. Change it based on your recommendation .

orelmisan commented on tests/instancetype/instancetype.go:
```diff
@@ -1743,7 +1760,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 								Domain: virtv1.DomainSpec{
 									Memory: &virtv1.Memory{
-										Guest: resource.NewQuantity(2*1024*1024*1024, resource.BinarySI),
+										Guest: resource.NewQuantity(memorySize2GB, resource.BinarySI),
```
Please consider using https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource#MustParse:
```suggestion
										Guest: resource.MustParse("2Gi"),
```
Same in other places defining memory quantity.

orelmisan commented on tests/instancetype/instancetype.go:
```diff
@@ -704,11 +710,12 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
+			preference              *instancetypev1beta1.VirtualMachinePreference
+			sourceDV                *cdiv1beta1.DataVolume
+			namespace               string
+			dvSuccessTimeoutSeconds = 180
```
Could this be a const?

orelmisan commented on tests/instancetype/instancetype.go:
```diff
@@ -297,7 +303,7 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 				PreferredUseBios: pointer.P(true),
 			}
-			preference.Spec.PreferredTerminationGracePeriodSeconds = pointer.P(int64(15))
+			preference.Spec.PreferredTerminationGracePeriodSeconds = pointer.P(int64(preferredTerminationGracePeriodSeconds))
```
KubeVirt has a generic pointer package:
https://github.com/kubevirt/kubevirt/blob/43531861ac5fc7795ac6523c482b2e2de74dfa3b/pkg/pointer/pointer.go#L22

Please consider using it.

orelmisan reviewed (commented):
Thank you for the changes @anishbista60.

anishbista60 commented:
@orelmisan  please let me know if there is anything to look into.

orelmisan commented:
> @orelmisan please let me know if there is anything to look into.

Thank you for the changes @anishbista60.
Could you please address https://github.com/kubevirt/kubevirt/pull/12694/files#r1741853940 ?
It was resolved without providing feedback for the concern that was raised.

anishbista60 commented:
> > @orelmisan please let me know if there is anything to look into.
> 
> Thank you for the changes @anishbista60. Could you please address https://github.com/kubevirt/kubevirt/pull/12694/files#r1741853940 ? It was resolved without providing feedback for the concern that was raised.

Regarding this, you told me to change `var` into `const`. But , when I raised this PR , I added  them as constant but later, on the recommendation of @lyarwood  I changed it to. `var`. Now, should I leave it as it is or change it. 

Thanykou

orelmisan commented:
> > > @orelmisan please let me know if there is anything to look into.
> > 
> > 
> > Thank you for the changes @anishbista60. Could you please address https://github.com/kubevirt/kubevirt/pull/12694/files#r1741853940 ? It was resolved without providing feedback for the concern that was raised.
> 
> Regarding this, you told me to change `var` into `const`. But , when I raised this PR , I added them as constant but later, on the recommendation of @lyarwood I changed it to. `var`. Now, should I leave it as it is or change it ?
> 
> Thank you

@lyarwood is the owner of this file, what he says has precedence.

orelmisan commented:
/test all

orelmisan commented on tests/instancetype/instancetype.go:
```diff
@@ -1422,9 +1429,19 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
+			preferCores    = instancetypev1beta1.Cores
+			preferThreads  = instancetypev1beta1.Threads
+			preferAny      = instancetypev1beta1.Any
+			guestMemory2GB = resource.MustParse("2GiB")
+			guestMemory1GB = resource.MustParse("1GiB")
```
@anishbista60 the tests fail due to a panic here:
```suggestion
			guestMemory2GB = resource.MustParse("2Gi")
			guestMemory1GB = resource.MustParse("1Gi")
```

lyarwood commented on tests/instancetype/instancetype.go:
```diff
@@ -1422,9 +1429,19 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
+		const (
+			providedvCPUs    = 2
+			requiredNumvCPUs = 2
+			requiredvCPUs    = 4
```
I still think the naming could be better here...

anishbista60 commented:
@lyarwood Done . 
this name is there. Based on your suggestion in previous review 

```
providedvCPUs    = 2
requiredNumvCPUs = 2
requiredvCPUs    = 4
```

lyarwood commented:
> @lyarwood Done . These name are present based on your suggestion in previous review
> 
> ```
> providedvCPUs    = 2
> requiredNumvCPUs = 2
> requiredvCPUs    = 4
> ```

Yup okay, for now this is good enough. Thanks for your contribution!

/approve

lyarwood commented:
> @lyarwood Done . These name are present based on your suggestion in previous review
> 
> ```
> providedvCPUs    = 2
> requiredNumvCPUs = 2
> requiredvCPUs    = 4
> ```

Yup okay, for now this is good enough. Thanks for your contribution!

/lgtm

(previously tried to /approve but missed this touched `.golangci.yml`)

anishbista60 commented:
@lyarwood linter test case throws an error 

```
tests/instancetype/instancetype.go:50:8: ST1011: var timeoutSeconds is of type time.Duration; don't use unit-specific suffix "Seconds" (stylecheck)
	const timeoutSeconds = 300 * time.Second
```
so , i change it to `timeout`

lyarwood commented:
/lgtm

TIL that's even a rule, odd that it's specific to `time.Duration` variables but I guess that makes sense.

anishbista60 commented:
@lyarwood  what is the reason of failing behind `pull-kubevirt-e2e-k8s-1.31-sig-compute` ?.

0xFelix commented:
It's a flaky test not related to your changes.

/retest-required

anishbista60 commented:
@lyarwood Could you please move further ?

anishbista60 commented:
> It's a flaky test not related to your changes.
> 
> /retest-required

Thankyou @0xFelix . I always remember your help at my First PR 7 months back. ❤️

anishbista60 commented:
anyone please look at this PR and try to get merge. Thannkyou

0xFelix reviewed (commented):
/test all

anishbista60 commented:
@lyarwood could you please move further ?

anishbista60 commented:
Please try to get it merge. only approval is needed. Thank you.

0xFelix reviewed (commented):
/approve
/test all

0xFelix reviewed (approved):
Not able to approve because `.golangi.yml` was changed.

/lgtm

anishbista60 commented:
@jean-edouard Please have a look .

jean-edouard commented:
/approve

anishbista60 commented:
Thank you everyone who helped in reviewing
